File: 算法笔记/NQSDemo/src/main/java/jeason/A_Leetcode/链表/_1669_合并两个有序链表_将一个链表插入替换到另一个链表的中间.java
package jeason.A_Leetcode.链表;
// https://leetcode.cn/problems/merge-in-between-linked-lists/


class _1669_合并两个有序链表_将一个链表插入替换到另一个链表的中间{

   //list1找到a、b两个节点，删除a-b中间的元素，将list2放进去

    //思路：遍历找到链表1保留部分的尾节点和链表2的尾节点。将两链表连接起来。

    public ListNode mergeInBetween(ListNode list1, int a, int b, ListNode list2) {
        //创建头部、中部、尾部的node节点
        ListNode headNode = list1, tempNode = list1, endNode = list2;
        
        int i = 0, j = 0;
        while(headNode != null && tempNode != null && j < b){
            //遍历找到入口a的节点，赋值给头部node
            if(i != a - 1){
                headNode = headNode.next;
                i++;
            } 
            //遍历找到出口b的节点，赋值给tempNode
            if(j != b){
                tempNode = tempNode.next;
                j++;
            } 
        }
        //-----------入口a连接到list2的头部----------
        headNode.next = list2;

        //-----------出口b连接到list2的尾部----------
        //tempMode移动到出口b的下一个节点
        tempNode = tempNode.next;
        //寻找list2的尾节点
        while(endNode.next != null){
            endNode = endNode.next;
        }
        endNode.next = tempNode;

        return list1;
    }
}

File: PythonAlgorithm/Leetcode/链表/_1669_合并两个有序链表_将一个链表插入替换到另一个链表的中间.py

# https://leetcode.cn/problems/merge-in-between-linked-lists/

""" 
给你两个链表 list1 和 list2 ，它们包含的元素分别为 n 个和 m 个。
请你将 list1 中下标从 a 到 b 的全部节点都删除，并将list2 接在被删除节点的位置。

输入：list1 = [0,1,2,33,444,55,6], a = 33, b = 55, list2 = [1001,1002,1003]
输出：[0,1,2,1001,1002,1003,6]
题解：
"""


class ListNode(object):
    def __init__(self, val: int = 0, next: 'ListNode' = None):
        self.val: int = val
        self.next: ListNode = next

class Solution(object):
    
    def mergeInBetween(self, list1: ListNode, indexOfStart: int, indexOfEnd: int, list2: ListNode) -> ListNode:
        
        # 找到需要替换区间的开头和结尾节点
        insertStartNode = insertEndNode = None
        
        tempNode: ListNode = list1
        i: int = 0
        while list1 and list1.next:
            # 插入的起始、结束点都结束了，中止循环
            if insertStartNode and insertEndNode:
                break
            
            # 找到起始点的前一个节点
            if (i == (indexOfStart - 1)) and (i < indexOfEnd):
                insertStartNode = tempNode
            # 找到结束点
            if (i == indexOfEnd):
                insertEndNode = tempNode
            
            i += 1
            tempNode = tempNode.next
            
        
        if insertEndNode and insertStartNode:
            # 起始点的前一个节点.next指向list2
            insertStartNode.next = list2
            # list2的最后一个节点.next指向->结束点.next
            tempInsertNode: ListNode = list2
            while tempInsertNode.next:
                tempInsertNode = tempInsertNode.next
            tempInsertNode.next = insertEndNode.next
            
        return list1


def run():
    list1: ListNode = generateLinkList(data=[1, 2, 33, 44, 55, 6])
    list2: ListNode = generateLinkList(data=[1000, 1001, 1002])
    
    result: ListNode = Solution().mergeInBetween(list1=list1, indexOfStart=2, indexOfEnd=4, list2=list2)
    print(f"result1={printLinkList(result)}")
    

# 生成单向链表
def generateLinkList(data: list) -> ListNode:
    header: ListNode = ListNode(666, None)
    lastNode: ListNode = header
    for i, ele in enumerate(data):
        tempNode: ListNode = ListNode(ele, None)
        lastNode.next = tempNode
        lastNode = tempNode
    return header.next

# 打印单向链表
def printLinkList(header: ListNode) -> str:
    
    printStr: str = f"{header.val}"
    while header.next is not None:
        header = header.next
        printStr = f"{printStr}, {header.val}"
    return f"{printStr}"

run()

